src/Butler/Auth/OIDC.hs

```diff
@@ -126,40 +126,42 @@ loginServer sessions mkIndexHtml jwtSettings oidcenv auth mWorkspace =
-        case (mErr, mCode, mState) of
-            (Just errorMsg, _, _) -> error $ "Error from remote provider: " <> show errorMsg
-            (_, Nothing, _) -> error "No code parameter given"
-            (_, _, Nothing) -> error "No state parameter given"
-            (_, Just oauthCode, Just oauthState) -> do
-                tokens :: O.Tokens Value <-
-                    liftIO $
-                        O.getValidTokens
-                            (mkSessionStore oidcenv (Just $ encodeUtf8 oauthState) Nothing)
-                            (oidc oidcenv)
-                            (manager oidcenv)
-                            (encodeUtf8 oauthState)
-                            (encodeUtf8 oauthCode)
-                let username = UserName tokens.idToken.sub
-                setSession username $ externalProvider "google" username
+    callbackRoute mErr mCode mState = withSession \case
+        Nothing -> guestRedirect -- this is weird, the user should already have a guest session at this point.
+        Just session ->
+            case (mErr, mCode, mState) of
+                (Just errorMsg, _, _) -> error $ "Error from remote provider: " <> show errorMsg
+                (_, Nothing, _) -> error "No code parameter given"
+                (_, _, Nothing) -> error "No state parameter given"
+                (_, Just oauthCode, Just oauthState) -> do
+                    tokens :: O.Tokens Value <-
+                        liftIO $
+                            O.getValidTokens
+                                (mkSessionStore oidcenv (Just $ encodeUtf8 oauthState) Nothing)
+                                (oidc oidcenv)
+                                (manager oidcenv)
+                                (encodeUtf8 oauthState)
+                                (encodeUtf8 oauthCode)
+                    let username = UserName tokens.idToken.sub
+                    setSession session username $ externalProvider "google" username
 
     guestCallbackRoute :: ServantProcessIO AuthResp
     guestCallbackRoute = do
         session <- lift $ newSession sessions Nothing "guest"
         setCookiesAndRedirect session
 
-    setSession :: UserName -> SessionProvider -> ServantProcessIO AuthResp
-    setSession username provider = do
+    setSession :: Session -> UserName -> SessionProvider -> ServantProcessIO AuthResp
+    setSession currentSession username provider = do
         let ctx = ["username" .= username, "provider" .= provider]
         session <-
             lift $
                 atomically (lookupSessionByProvider sessions provider) >>= \case
                     Nothing -> do
-                        logInfo "Creating a new session" ctx
-                        newSession sessions (Just provider) username
+                        logInfo "Promoting current session to the new provider" ctx
+                        changeProvider sessions currentSession provider
+                        pure currentSession
```

Comment:
I guess we could skip setting the cookie in that case, and directly redirect to the indexRoute.

Reply:
We could but if we want to map the external IDP session validity we received via the JWT to the cookie we set to the client then we should keep setting the cookie I think.